---
title: "Merge 10x Genomics scRNA-seq .h5 files generated by split_h5_by_hash.Rmd"
author: 
 - Lucas Graybuck
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    df_print: paged
    self_contained: true
params:
  in_dir: NULL
  out_dir: NULL
---

```{r Load Libraries}
start_time <- Sys.time()

quiet_library <- function(...) {
  suppressPackageStartupMessages(library(...))
}
quiet_library(rhdf5)
quiet_library(H5weaver)
quiet_library(Matrix)
quiet_library(ggplot2)
quiet_library(cowplot)
```

Declaring start
```{r Declare start}
stm("Merging .h5 files")
```

Argument parsing
```{r Parse arguments}
if(is.null(params$in_dir)) {
  in_dir <- system.file("testdata/splitdata", package = "H5weaver")
  out_dir <- tempdir()
} else {
  in_dir <- params$in_dir
  out_dir <- params$out_dir
}

stm(paste0("IN  H5 directory    : ", in_dir))
stm(paste0("OUT Directory       : ", out_dir))
```

```{r Print Arguments}
print(c(
  paste0("IN  H5 directory    : ", in_dir),
  paste0("OUT Directory       : ", out_dir)
))
```

Locate input files
```{r Locate inputs}
in_file_names <- list.files(in_dir, 
                            pattern = "_.+h5$")
# Check for per-well directories
if(length(in_file_names) == 0) {
  in_file_dirs <- list.dirs(in_dir, 
                            recursive = FALSE)
  in_file_names <- unlist(lapply(in_file_dirs,
                                 list.files, 
                                 pattern = "_.+h5$"))
  in_file_paths <- unlist(lapply(in_file_dirs,
                                 list.files,
                                 pattern = "_.+h5$", 
                                 full.names = TRUE))
} else {
  in_file_paths <- list.files(in_dir, 
                              pattern = "_.+h5$", 
                              full.names = TRUE)
}

stm(paste0("Found ", length(in_file_paths), " .h5 files for processing"))
print(paste0("Found ", length(in_file_paths), " .h5 files for processing"))
```

Identify hashes in in_dir
```{r Identify hashes}
in_file_hashes <- sub(".h5$","",sub(".+_","",in_file_names))
unique_hashes <- unique(in_file_hashes)

stm(paste0("Identified ", length(unique_hashes), " hash patterns: ", paste(unique_hashes, collapse = ", ")))
```

Merge files based on hashes
```{r Load Count Matrix}
meta_list <- list()
for(i in seq_along(unique_hashes)) {
  current_hash <- unique_hashes[i]
  stm(paste0("Merging for hash ", current_hash))
  current_files <- in_file_paths[in_file_hashes == current_hash]
  for(f in current_files) {
    stm(paste0("IN  H5 file         : ", f))
  }
  
  h5_ll <- lapply(current_files,
                  h5dump)
  
  h5_list <- reduce_h5_list(h5_ll,
                            sparse_matrices = c("matrix","hash"))
  
  meta_list[[current_hash]] <- h5_list_cell_metadata(h5_list)
  
  out_file <- file.path(out_dir, paste0(current_hash, ".h5"))
  stm(paste0("Writing HDF5 to     : ", out_file))
  
  write_h5_list(h5_list,
                out_file,
                overwrite = TRUE)
  
  # transposed version for visualization
  h5_list <- h5_list_transpose(h5_list,
                               sparse_matrices = c("matrix","hash"))
  t_out_file <- file.path(out_dir, paste0("transposed_",current_hash, ".h5"))
  stm(paste0("Writing HDF5 to     : ", t_out_file))
  
  write_h5_list(h5_list,
                t_out_file,
                overwrite = TRUE)
} 

all_meta <- do.call("rbind", meta_list)
all_meta <- as.data.table(all_meta)
```

```{r Declare Metrics and Plotting}
stm("Generating QC Metrics and Plots for Report")
```

```{r HTO Category Count By Well}
hto_category_by_well <- all_meta[, 
                                 .(n_cells = nrow(.SD)),
                                 by = list(well_id, hto_category)]

hto_category_count_by_well <- dcast(hto_category_by_well,
                                    formula = well_id ~ hto_category,
                                    value.var = "n_cells")
knitr::kable(hto_category_count_by_well,
             caption = "HTO Category Count per Well")
```

```{r HTO Category Count By Well Plot, fig.height = 6, fig.width = 10}
qc_aligned_barplot(all_meta,
                   category_x = "well_id",
                   name_x = "Well ID",
                   category_y = "hto_category",
                   category_name = "HTO Category",
                   colorset_y = "varibow",
                   name_y = "N Cells",
                   padding = 0.2) +
  ggtitle("HTO Category Counts per Well")
```

```{r HTO Category Fraction By Well}
hto_category_by_well <- hto_category_by_well[,
                                             frac_cells := round(n_cells/sum(n_cells),4),
                                             by = well_id]

hto_category_frac_by_well <- dcast(hto_category_by_well,
                                   formula = well_id ~ hto_category,
                                   value.var = "frac_cells")

knitr::kable(hto_category_frac_by_well,
             caption = "HTO Category Fraction per Well")
```

```{r HTO Category Fraction By Well Plot, fig.height = 6, fig.width = 10}
qc_stacked_barplot(all_meta,
                   category_x = "well_id",
                   name_x = "Well ID",
                   category_y = "hto_category",
                   category_name = "HTO Category",
                   colorset_y = "varibow",
                   name_y = "Fraction of Cells",
                   as_fraction = TRUE) +
  ggtitle("HTO Category Fraction per Well")
```

Plot distribution of samples per well and wells per sample:
```{r Filter Singlets}
singlet_meta <- all_meta[hto_category == "singlet"]
```

```{r HTO Barcode Count By Well}
hto_barcode_by_well <- singlet_meta[, 
                                    .(n_cells = nrow(.SD)),
                                    by = list(well_id, hto_barcode)]

hto_barcode_count_by_well <- dcast(hto_barcode_by_well,
                                   formula = well_id ~ hto_barcode,
                                   value.var = "n_cells")
knitr::kable(hto_barcode_count_by_well,
             caption = "Singlet HTO Barcode Count per Well")
```

```{r HTO Barcode Count By Well Plot, fig.height = 6, fig.width = 10}
qc_aligned_barplot(singlet_meta,
                   category_x = "well_id",
                   name_x = "Well ID",
                   category_y = "hto_barcode",
                   category_name = "HTO Barcode",
                   colorset_y = "varibow",
                   name_y = "N Cells",
                   padding = 0.2) +
  ggtitle("HTO Barcode Distribution in Wells")
```
```{r Wells by HTO Barcode Count Plot, fig.height = 6, fig.width = 10}
qc_aligned_barplot(singlet_meta,
                   category_x = "hto_barcode",
                   name_x = "HTO Barcode",
                   category_y = "well_id",
                   category_name = "Well ID",
                   colorset_y = "varibow",
                   name_y = "N Cells",
                   padding = 0.2) +
  ggtitle("Well Distribution in HTO Barcodes")
```
```{r HTO Barcode Fraction By Well}
hto_barcode_by_well <- hto_barcode_by_well[,
                                           frac_cells := round(n_cells/sum(n_cells),4),
                                           by = well_id]

hto_barcode_frac_by_well <- dcast(hto_barcode_by_well,
                                  formula = well_id ~ hto_barcode,
                                  value.var = "frac_cells")

knitr::kable(hto_barcode_frac_by_well,
             caption = "HTO Barcode Fraction per Well")
```

```{r HTO Barcode Fraction By Well Plot, fig.height = 6, fig.width = 10}
qc_stacked_barplot(singlet_meta,
                   category_x = "well_id",
                   name_x = "Well ID",
                   category_y = "hto_barcode",
                   category_name = "HTO Barcode",
                   colorset_y = "varibow",
                   name_y = "Fraction of Cells",
                   as_fraction = TRUE) +
  ggtitle("HTO Barcode Fraction per Well")
```

```{r Well Fraction By HTO Barcode Plot, fig.height = 6, fig.width = 10}
qc_stacked_barplot(singlet_meta,
                   category_x = "hto_barcode",
                   category_y = "well_id",
                   category_name = "Well ID",
                   name_x = "HTO Barcode",
                   colorset_y = "varibow",
                   name_y = "Fraction of Cells",
                   as_fraction = TRUE) +
  ggtitle("Well Fraction per HTO Barcode")
```

Counts, Median Reads, UMIs, Genes across all wells
```{r}
overview_metrics <- all_meta[,.(n_cells = nrow(.SD),
                                median_reads = median(n_reads),
                                median_umis = median(n_umis),
                                median_genes = median(n_genes))]

knitr::kable(overview_metrics,
             caption = "Run Metrics Overview")
```

Median Reads, UMIs, Genes per well
```{r}
well_metrics <- all_meta[,.(n_cells = nrow(.SD),
                            median_reads = as.numeric(median(n_reads)),
                            median_umis = as.numeric(median(n_umis)),
                            median_genes = as.numeric(median(n_genes))),
                         by = well_id]

knitr::kable(well_metrics,
             caption = "Well Metrics Overview")
```

```{r}
qc_violin_plot(all_meta,
               category_x = "well_id",
               name_x = "Well ID",
               column_y = "n_reads",
               name_y = "N Reads per Cell",
               fill = "dodgerblue") +
  ggtitle("Reads per Cell")

```

```{r}
qc_violin_plot(all_meta,
               category_x = "well_id",
               name_x = "Well ID",
               column_y = "n_umis",
               name_y = "N UMIs per Cell",
               fill = "purple") +
  ggtitle("UMIs per Cell")

```

```{r}
qc_violin_plot(all_meta,
               category_x = "well_id",
               name_x = "Well ID",
               column_y = "n_genes",
               name_y = "N Genes per Cell",
               fill = "orangered") +
  ggtitle("Genes per Cell")

```

Median Reads, UMIs, Genes per HTO Barcode
```{r}
hto_barcode_metrics <- singlet_meta[,.(n_cells = nrow(.SD),
                                       median_reads = as.numeric(median(n_reads)),
                                       median_umis = as.numeric(median(n_umis)),
                                       median_genes = as.numeric(median(n_genes))),
                                    by = hto_barcode]

knitr::kable(well_metrics,
             caption = "HTO Barcode Metrics Overview")
```

```{r}
category_reads_violins <- qc_violin_plot(all_meta,
                                         category_x = "hto_category",
                                         name_x = "HTO Category",
                                         column_y = "n_reads",
                                         name_y = "N Reads per Cell",
                                         fill = "dodgerblue")

barcode_reads_violins <- qc_violin_plot(all_meta[all_meta$hto_category == "singlet",],
                                        category_x = "hto_barcode",
                                        name_x = "HTO Barcode (singlets)",
                                        column_y = "n_reads",
                                        name_y = "N Reads per Cell",
                                        fill = "dodgerblue") +
  ggtitle("Reads per Cell")

reads_violin_list <- list(category_reads_violins, 
                          barcode_reads_violins)

plot_grid(plotlist = reads_violin_list,
          ncol = 2, rel_widths = c(1, 3),
          nrow = 1, align = "h")
```

```{r}
category_umis_violins <- qc_violin_plot(all_meta,
                                        category_x = "hto_category",
                                        name_x = "HTO Category",
                                        column_y = "n_umis",
                                        name_y = "N UMIs per Cell",
                                        fill = "purple")

barcode_umis_violins <- qc_violin_plot(all_meta[all_meta$hto_category == "singlet",],
                                       category_x = "hto_barcode",
                                       name_x = "HTO Barcode (singlets)",
                                       column_y = "n_umis",
                                       name_y = "N UMIs per Cell",
                                       fill = "purple") +
  ggtitle("UMIs per Cell")

umis_violin_list <- list(category_umis_violins, 
                         barcode_umis_violins)

plot_grid(plotlist = umis_violin_list,
          ncol = 2, rel_widths = c(1, 3),
          nrow = 1, align = "h")
```

```{r}
category_genes_violins <- qc_violin_plot(all_meta,
                                         category_x = "hto_category",
                                         name_x = "HTO Category",
                                         column_y = "n_genes",
                                         name_y = "N Genes per Cell",
                                         fill = "orangered")

barcode_genes_violins <- qc_violin_plot(all_meta[all_meta$hto_category == "singlet",],
                                        category_x = "hto_barcode",
                                        name_x = "HTO Barcode (singlets)",
                                        column_y = "n_genes",
                                        name_y = "N Genes per Cell",
                                        fill = "orangered") +
  ggtitle("Genes per Cell")

genes_violin_list <- list(category_genes_violins, 
                          barcode_genes_violins)

plot_grid(plotlist = genes_violin_list,
          ncol = 2, rel_widths = c(1, 3),
          nrow = 1, align = "h")
```

Median Reads per Well By HTO Barcode
```{r}
reads_hto_barcode_by_well <- singlet_meta[,
                                          .(median_reads = median(n_reads)),
                                          by = list(well_id, hto_barcode)]

reads_hto_barcode_by_well <- dcast(reads_hto_barcode_by_well,
                                   formula = well_id ~ hto_barcode,
                                   value.var = "median_reads")

knitr::kable(reads_hto_barcode_by_well,
             caption = "Median Reads per Singlet HTO Barcode Fraction By Well")
```

Median UMIs per Well By HTO Barcode
```{r}
umis_hto_barcode_by_well <- singlet_meta[,
                                         .(median_umis = median(n_umis)),
                                         by = list(well_id, hto_barcode)]

umis_hto_barcode_by_well <- dcast(umis_hto_barcode_by_well,
                                  formula = well_id ~ hto_barcode,
                                  value.var = "median_umis")

knitr::kable(umis_hto_barcode_by_well,
             caption = "Median UMIs per Singlet HTO Barcode Fraction By Well")
```

Median Genes per Well By HTO Barcode
```{r}
genes_hto_barcode_by_well <- singlet_meta[,
                                          .(median_genes = median(n_genes)),
                                          by = list(well_id, hto_barcode)]

genes_hto_barcode_by_well <- dcast(genes_hto_barcode_by_well,
                                   formula = well_id ~ hto_barcode,
                                   value.var = "median_genes")

knitr::kable(genes_hto_barcode_by_well,
             caption = "Median Genes per Singlet HTO Barcode Fraction By Well")
```

```{r Session Info}
sessionInfo()
```

Total time elapsed
```{r Show Time}
end_time <- Sys.time()
diff_time <- end_time - start_time
time_message <- paste0("Elapsed Time: ", 
                       round(diff_time, 3),
                       " ", units(diff_time))
print(time_message)
stm(time_message)
stm("H5 merge process complete.")
```
