---
title: "Split 10x Genomics scRNA-seq .h5 file based on HTOparser results"
author: 
 - Lucas Graybuck
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    df_print: paged
    self_contained: true
params:
  in_h5: NULL
  in_mol: NULL
  in_mat: NULL
  in_tbl: NULL
  in_well: NULL
  out_dir: NULL
---

```{r Load Libraries}
start_time <- Sys.time()

quiet_library <- function(...) {
  suppressPackageStartupMessages(library(...))
}
quiet_library(rhdf5)
quiet_library(H5weaver)
quiet_library(Matrix)
quiet_library(ggplot2)
quiet_library(cowplot)
```

Declaring start
```{r Declare start}
stm("Starting .h5 split")
```

Argument parsing
```{r Parse arguments}
if(is.null(params$in_h5)) {
  in_h5 <- system.file("testdata/well1.h5", package = "H5weaver")
  in_mol <- system.file("testdata/well1_molecule_info.h5", package = "H5weaver")
  in_mat <- system.file("testdata/well1_count_matrix.csv.gz", package = "H5weaver")
  in_tbl <- system.file("testdata/well1_category_table.csv.gz", package = "H5weaver")
  in_well <- "B000-P0C0W0"
  out_dir <- tempdir()
} else {
  in_h5 <- params$in_h5
  in_mol <- params$in_mol
  in_mat <- params$in_mat
  in_tbl <- params$in_tbl
  in_well <- params$in_well
  out_dir <- params$out_dir
}

stm(paste0("IN  H5 file         : ", in_h5))
stm(paste0("IN  Mol Info H5     : ", in_mol))
stm(paste0("IN  Count matrix    : ", in_mat))
stm(paste0("IN  Category table  : ", in_tbl))
stm(paste0("IN  Well ID         : ", in_well))
stm(paste0("OUT Directory       : ", out_dir))
```

```{r Print Arguments}
print(c(
  paste0("IN  H5 file         : ", in_h5),
  paste0("IN  Mol Info H5     : ", in_mol),
  paste0("IN  Count matrix    : ", in_mat),
  paste0("IN  Category table  : ", in_tbl),
  paste0("IN  Well ID         : ", in_well),
  paste0("OUT Directory       : ", out_dir)
))
```

Load HTO counts
```{r Load Count Matrix}
stm(paste0("Loading HTO count matrix from ", in_mat))
hash_count_df <- fread(in_mat)

hash_count_mat <- as.matrix(hash_count_df[,-1])

rownames(hash_count_mat) <- hash_count_df[[1]]
rm(hash_count_df)
```

Load HTO categories
```{r Load Category Table}
stm(paste0("Loading HTO category table from ", in_tbl))
hash_category_table <- fread(in_tbl)

hto_barcodes <- unique(hash_category_table$hto_barcode[hash_category_table$hto_category == "singlet"])
```

Display expected output files
```{r Display Outputs}
output_h5_files <- file.path(out_dir, paste0(in_well, "_", 
                                             c(hto_barcodes,"multiplet"),
                                             ".h5"))

output_messages <- paste0("OUT HDF5            : ", output_h5_files)

for(i in seq_along(output_messages)) {
  stm(output_messages[i])
}

print(output_messages)
```

Load scRNA_seq  Dataset
```{r Load Query}
stm(paste0("Loading HDF5 from ", in_h5))
h5_list <- h5dump(in_h5)
```

Read molecule info to get read counts per cell
```{r Assemble Read Counts}
stm(paste0("Assembling Read Counts per Cell from ", in_mol))

bc <- sub("-1","",h5_list$matrix$barcodes)

bc_counts <- data.table(mol_idx = h5read(in_mol, "/barcode_idx"),
                        umi_count = h5read(in_mol, "/count"))

bc_sums <- bc_counts[, .(n_reads = sum(umi_count)), by = mol_idx]
rm(bc_counts)

mol_bc <- h5read(in_mol, "/barcodes")
bc_sums$cell_barcode <- mol_bc[bc_sums$mol_idx + 1]
rm(mol_bc)

bc_sums <- bc_sums[,.(cell_barcode, n_reads)]

n_reads <- bc_sums$n_reads[match(bc, bc_sums$cell_barcode)]
n_reads[is.na(n_reads)] <- 0

if("observations" %in% names(h5_list$matrix)) {
  h5_list$matrix$observations$n_reads <- n_reads
} else {
  h5_list$matrix$observations <- list(n_reads = n_reads)
}
```

Compute N UMIs and N Genes per cell
```{r n_umi and n_genes}
stm("Computing UMI and Gene Counts per Cell")

h5_list <- h5_list_convert_to_dgCMatrix(h5_list, target = "matrix")
h5_list$matrix$observations$n_umis <- unname(colSums(h5_list$matrix_dgCMatrix))
h5_list$matrix$observations$n_genes <- unname(colSums(h5_list$matrix_dgCMatrix > 0))

h5_list <- h5_list_convert_from_dgCMatrix(h5_list, target = "matrix")
```

Add cell ids
```{r}
stm("Adding Cell UUIDs and Names")

h5_list <-add_cell_ids(h5_list,
                       add_uuid = TRUE,
                       replace_barcode = TRUE,
                       retain_original_barcode = TRUE,
                       add_name = TRUE)
```

Generate Chip, Pool, and Batch IDs based on well_id
```{r Chip Pool Batch}
stm("Adding Batch, Pool, Chip, and Well metadata")

h5_list <- add_well_metadata(h5_list,
                             well_id = in_well)
```

Add hash data to the list object
```{r Add Hash Data}
stm("Adding hash data to cellranger results")

h5_list <- add_h5_list_hash_results(h5_list,
                                    hash_category_table,
                                    hash_count_mat,
                                    match_target = "original_barcodes")
```

Extract metadata for plotting
```{r Assemble Metadata}
meta <- h5_list_cell_metadata(h5_list)
```


Plot read/umi/gene histograms
```{r QC histograms}
reads_plot <- qc_hist_plot(meta,
                           column = "n_reads",
                           name_x = "N Reads per Cell",
                           fill = "dodgerblue",
                           target = 2e4)

umis_plot <- qc_hist_plot(meta,
                           column = "n_umis",
                           name_x = "N UMIs per Cell",
                           fill = "purple",
                           target = 8e3)

genes_plot <- qc_hist_plot(meta,
                           column = "n_genes",
                           name_x = "N Genes per Cell",
                           fill = "orangered",
                           target = 2e3)

histogram_list <- list(reads_plot,
                       umis_plot,
                       genes_plot)

plot_grid(plotlist = histogram_list,
          ncol = 1,
          nrow = 3)
```

Plot read/umi/gene scatters
```{r}
qc_scatter_plot(meta,
                column_x = "n_reads",
                name_x = "N Reads per Cell",
                column_y = "n_umis",
                name_y = "N UMIs per Cell",
                color = "darkblue") +
  ggtitle("UMIs by Reads")

qc_scatter_plot(meta,
                column_x = "n_umis",
                name_x = "N UMIs per Cell",
                column_y = "n_genes",
                name_y = "N Genes per Cell",
                color = "red") +
  ggtitle("Genes by UMIs")

qc_scatter_plot(meta,
                column_x = "n_reads",
                name_x = "N Reads per Cell",
                column_y = "n_genes",
                name_y = "N Genes per Cell",
                color = "darkgreen") +
  ggtitle("Genes by Reads")
```

Read distributions by HTO category and HTO barcode
```{r}
category_reads_violins <- qc_violin_plot(meta,
                                         category_x = "hto_category",
                                         name_x = "HTO Category",
                                         column_y = "n_reads",
                                         name_y = "N Reads per Cell",
                                         fill = "dodgerblue")

barcode_reads_violins <- qc_violin_plot(meta[meta$hto_category == "singlet",],
                                        category_x = "hto_barcode",
                                        name_x = "HTO Barcode (singlets)",
                                        column_y = "n_reads",
                                        name_y = "N Reads per Cell",
                                        fill = "dodgerblue") +
  ggtitle("Reads per Cell")

reads_violin_list <- list(category_reads_violins, 
                          barcode_reads_violins)

plot_grid(plotlist = reads_violin_list,
          ncol = 2, rel_widths = c(1, 3),
          nrow = 1, align = "h")
```

UMI distributions by HTO category and HTO barcode
```{r}
category_umis_violins <- qc_violin_plot(meta,
                                         category_x = "hto_category",
                                         name_x = "HTO Category",
                                         column_y = "n_umis",
                                         name_y = "N UMIs per Cell",
                                         fill = "purple")

barcode_umis_violins <- qc_violin_plot(meta[meta$hto_category == "singlet",],
                                        category_x = "hto_barcode",
                                        name_x = "HTO Barcode (singlets)",
                                        column_y = "n_umis",
                                        name_y = "N UMIs per Cell",
                                        fill = "purple") +
  ggtitle("UMIs per Cell")

umis_violin_list <- list(category_umis_violins, 
                          barcode_umis_violins)

plot_grid(plotlist = umis_violin_list,
          ncol = 2, rel_widths = c(1, 3),
          nrow = 1, align = "h")
```

Gene distributions by HTO category and HTO barcode
```{r}
category_genes_violins <- qc_violin_plot(meta,
                                         category_x = "hto_category",
                                         name_x = "HTO Category",
                                         column_y = "n_genes",
                                         name_y = "N Genes per Cell",
                                         fill = "orangered")

barcode_genes_violins <- qc_violin_plot(meta[meta$hto_category == "singlet",],
                                        category_x = "hto_barcode",
                                        name_x = "HTO Barcode (singlets)",
                                        column_y = "n_genes",
                                        name_y = "N Genes per Cell",
                                        fill = "orangered") +
  ggtitle("Genes per Cell")

genes_violin_list <- list(category_genes_violins, 
                          barcode_genes_violins)

plot_grid(plotlist = genes_violin_list,
          ncol = 2, rel_widths = c(1, 3),
          nrow = 1, align = "h")
```

Split list based on hashes
```{r Split H5 Lists}
stm("Splitting HDF5 based on Hash Category Table")
h5_list <- split_h5_list_by_hash(h5_list)
```

Write HDF5 files
```{r Write split files}
for(i in seq_along(h5_list)) {
  out_file <- file.path(out_dir, paste0(in_well, "_", names(h5_list)[i],".h5"))
  stm(paste0("Writing HDF5 to ", out_file))
  write_h5_list(h5_list[[i]],
                h5_file = out_file)
  h5closeAll()
}
```

```{r Session Info}
sessionInfo()
```

Total time elapsed
```{r Show Time}
end_time <- Sys.time()
diff_time <- end_time - start_time
time_message <- paste0("Elapsed Time: ", 
                       round(diff_time, 3),
                       " ", units(diff_time))
print(time_message)
stm(time_message)
stm("HTO count processing complete.")
```
